### PR TITLE
[reward, cfg] feat: add multi-reward function support with weighted aggregation

### PR DESCRIPTION
**Pull request by ruihanglix**
### What does this PR do?

Adds support for composing multiple reward functions with weighted aggregation during RL training. Previously, only a single `custom_reward_function` could be specified. This PR introduces a `reward_functions` dict-based config that allows users to define multiple named reward functions, each with its own weight, and aggregate them via weighted sum.

This enables multi-objective RL scenarios (e.g., simultaneously optimizing OCR accuracy and image quality) without requiring users to manually combine rewards in a single monolithic function.

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: https://github.com/verl-project/verl-omni/pulls?q=is%3Apr+multi+reward
- [x] Format the PR title as `[{modules}] {type}: {description}`

### Test

Validated on 4x NVIDIA B300 GPUs with the Qwen-Image OCR FlowGRPO recipe, combining two reward functions:
- **OCR reward** (`genrm_ocr.py`): GRM-based OCR accuracy score via Qwen3-VL-8B-Instruct, weight=0.5
- **JPEG compressibility** (`jpeg_compressibility.py`): rewards compressible images (penalizes high-frequency noise/artifacts), weight=0.5

Training ran for 300 steps with `batch_size=32`, `n=16`, `lr=3e-5`, `image_resolution=512x512`.

**Results (training):**

| Metric | Step 1 | Step 25 | Step 50 | Step 100 | Step 150 | Step 200 | Step 250 | Step 300 |
|--------|--------|---------|---------|----------|----------|----------|----------|----------|
| `reward/ocr/mean` | 0.12 | 0.26 | 0.62 | 0.93 | 0.94 | 0.96 | 0.96 | 0.95 |
| `reward/jpeg/mean` | -0.22 | -0.21 | -0.17 | -0.14 | -0.13 | -0.11 | -0.10 | -0.08 |
| `reward/combined/mean` | -0.05 | 0.03 | 0.22 | 0.40 | 0.40 | 0.42 | 0.43 | 0.43 |

**Validation (step 300):** `reward/ocr/mean@1 = 0.98`, `reward/jpeg/mean@1 = -0.07`, `reward/combined/mean@1 = 0.45`

**Training curves:**

image

The OCR reward converges rapidly (~50 steps to 0.6, ~100 steps to 0.93), while the JPEG compressibility reward improves steadily throughout training (-0.22 → -0.08). This demonstrates that the model successfully optimizes both objectives simultaneously without one dominating the other.

**Qualitative examples (validation generations):**

Prompt 1: *"A birthday card interior with a whimsical, pastel-colored design, featuring the message 'Make A Wish' in elegant, cursive handwriting, surrounded by sparkling candles and colorful confetti."*

| Step 30 | Step 60 | Step 90 | Step 120 | ... | Step 300 |
|---------|---------|---------|----------|-----|----------|
| image | image | image | image | ... | image |
| score: 0.43 | score: 0.45 | score: 0.46 | score: 0.46 | | score: 0.47 |

Prompt 2: *"A wooden trail marker stands in a dense forest, its surface weathered by time. The words 'Hidden Trail' are carved deeply into the wood."*

| Step 30 | Step 60 | Step 90 | Step 120 | ... | Step 300 |
|---------|---------|---------|----------|-----|----------|
| image | image | image | image | ... | image |
| score: -0.11 | score: 0.43 | score: 0.43 | score: 0.44 | | score: 0.46 |

Key observations:
- **Text rendering is accurate from early training** — both prompts produce legible text by step 30, and text clarity remains stable throughout.
- **Image composition and detail improve** — later steps show more refined layouts, better color consistency, and richer scene details (e.g., candle flames, wood texture).
- **JPEG reward encourages compressibility** — the JPEG reward drives the model toward smoother, more compressible outputs (less high-frequency noise/artifacts). This is visible as backgrounds becoming softer at later steps, demonstrating the multi-objective trade-off in action.

### API and Usage Example

Multi-reward is configured via the `reward.reward_functions` dict. Each entry specifies a `path`, `name`, and `weight`:

```bash
python3 -m verl_omni.trainer.main_diffusion \
    # ... other args ...
    "+reward.reward_functions.ocr.path=verl_omni/utils/reward_score/genrm_ocr.py" \
    "+reward.reward_functions.ocr.name=compute_score_ocr" \
    "+reward.reward_functions.ocr.weight=0.5" \
    "+reward.reward_functions.jpeg.path=verl_omni/utils/reward_score/jpeg_compressibility.py" \
    "+reward.reward_functions.jpeg.name=compute_score" \
    "+reward.reward_functions.jpeg.weight=0.5" \
    reward.aggregation=weighted_sum
```

Key points:
- `reward.reward_functions` is mutually exclusive with `reward.custom_reward_function.path` — using both raises a `ValueError`.
- Each sub-reward function is called with filtered kwargs matching its signature (no need to accept unused parameters).
- Per-sub-reward scores are logged under `critic/reward/{key}/mean` in wandb for independent monitoring.
- Async reward functions are natively supported.

A full example script is provided at `examples/flowgrpo_trainer/run_qwen_image_ocr_multi_reward.sh`.

#### High-level architecture

```
┌─────────────────────────────────────────────────────────────────┐
│                    main_diffusion.py (entrypoint)                │
│                                                                 │
│  resolve_multi_reward_config(config)                            │
│    ├─ reward_functions is empty? → no-op (backward compat)      │
│    └─ reward_functions is set?   → rewrite config to wire up    │
│         MultiVisualRewardManager as the reward_manager class    │
└────────────────────────────────────┬────────────────────────────┘
                                     │
                                     ▼
┌─────────────────────────────────────────────────────────────────┐
│              MultiVisualRewardManager (extends VisualRM)         │
│                                                                 │
│  __init__:                                                      │
│    for each entry in reward_functions:                           │
│      load_extern_object(path, name) → fn                        │
│      inspect.signature(fn) → sig                                │
│      store {key, fn, weight, sig, is_async}                     │
│                                                                 │
│  run_single(data):                                              │
│    build all_kwargs from DataProto                               │
│    for each sub-reward:                                         │
│      filter_kwargs(all_kwargs, sig)  ← only pass declared args  │
│      call fn (async or via executor)                            │
│      accumulate: combined += weight * score                     │
│    return {reward_score, reward_extra_info}                     │
└────────────────────────────────────┬────────────────────────────┘
                                     │
                                     ▼
┌─────────────────────────────────────────────────────────────────┐
│          ray_diffusion_trainer.py (metric logging)              │
│                                                                 │
│  Iterates reward_extra_infos_dict to log per-sub-reward means: │
│    critic/reward/ocr/mean, critic/reward/jpeg/mean, etc.        │
└─────────────────────────────────────────────────────────────────┘
```

#### Design decisions

1. **Config-level wiring (not code-level branching):** Rather than adding `if multi_reward:` branches throughout the codebase, `resolve_multi_reward_config()` rewrites the OmegaConf config at startup to point `reward_manager.name` at `MultiVisualRewardManager`. The rest of the system (worker init, Ray resource pools) remains unaware of multi-reward. This keeps the blast radius minimal.

2. **Signature-based kwarg filtering:** Different reward functions need different inputs (e.g., OCR needs `reward_router_address` for the VLM, JPEG only needs `solution_image`). Instead of requiring all functions to accept `**kwargs`, we introspect each function's signature and only pass the parameters it declares. Functions that *do* declare `**kwargs` receive everything.

3. **Single reward model instance:** All sub-reward functions that require a reward model (VLM) share the same `reward_router_address`. This avoids GPU memory duplication. If a future use case requires multiple distinct models, the architecture can be extended by routing per-function.

4. **Graceful error handling:** If a sub-reward raises an exception, it contributes 0 to the weighted sum and logs an error, rather than crashing the entire training loop. This is important for robustness when one reward depends on an external service (e.g., VLM inference).

5. **Mutual exclusivity with `custom_reward_function`:** Using both `reward_functions` and `custom_reward_function.path` simultaneously is explicitly forbidden via a `ValueError`. This prevents ambiguous behavior.

#### File-by-file changes

| File | Change |
|------|--------|
| `verl_omni/reward_loop/reward_manager/multi.py` | New. Core `MultiVisualRewardManager` class with weighted aggregation logic. |
| `verl_omni/reward_loop/config.py` | New. `resolve_multi_reward_config()` for transparent config rewriting. |
| `verl_omni/reward_loop/reward_manager/__init__.py` | Export `MultiVisualRewardManager`. |
| `verl_omni/trainer/config/reward/reward.yaml` | Add `reward_functions: {}` and `aggregation: weighted_sum` defaults. |
| `verl_omni/trainer/main_diffusion.py` | Call `resolve_multi_reward_config(config)` after OmegaConf resolution. |
| `verl_omni/trainer/diffusion/ray_diffusion_trainer.py` | Log per-sub-reward means from `reward_extra_info` dict. |
| `examples/flowgrpo_trainer/run_qwen_image_ocr_multi_reward.sh` | New. Full example script combining OCR + JPEG rewards. |

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/verl-project/verl-omni/blob/main/CONTRIBUTING.md).
- [x] Apply pre-commit checks.
- [x] Backward compatible — existing single-reward configs work unchanged (`reward_functions` defaults to `{}`).
- [x] Validated with end-to-end training experiment (300 steps, 4xB300).

### CODE REVIEW

File: verl_omni/reward_loop/config.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        )
+
+    # Mutate config to wire up MultiVisualRewardManager
+    with open_dict(config):
+        # Set placeholder so upstream RewardLoopWorker loads something (never called directly)
+        config.reward.custom_reward_function.path = "pkg://verl_omni.reward_loop.reward_manager.multi"
+        config.reward.custom_reward_function.name = "_multi_reward_placeholder"
+        # Point reward_manager to MultiVisualRewardManager
+        config.reward.reward_manager.name = "MultiVisualRewardManager"
+        config.reward.reward_manager.module.path = "pkg://verl_omni.reward_loop.reward_manager"
```
Comment (zhtmike) [inline verl_omni/reward_loop/config.py]: suggest to explicitly apply these config in shell script. This wiring way is not obvious and clear to the user

File: verl_omni/trainer/config/reward/reward.yaml [hunk 2]
```diff
@@ -37,7 +45,7 @@ reward_model:
     name: ???
     dtype: bfloat16
     gpu_memory_utilization: 0.5
-    enforce_eager: true
+    enforce_eager: false
```
Comment (zhtmike) [inline verl_omni/trainer/config/reward/reward.yaml]: please do not change the default value unless you have provided the reason.

File: verl_omni/trainer/diffusion/ray_diffusion_trainer.py [hunk 3]
```diff
@@ -1097,6 +1097,17 @@ def fit(self):
...
+                if reward_extra_infos_dict:
+                    for key, values in reward_extra_infos_dict.items():
+                        if isinstance(values, np.ndarray):
+                            if not np.issubdtype(values.dtype, np.number):
+                                continue
+                            metrics[f"critic/{key}/mean"] = float(values.mean())
+                        elif isinstance(values, list) and len(values) > 0:
+                            if not isinstance(values[0], (int, float)):
+                                continue
+                            metrics[f"critic/{key}/mean"] = float(np.mean(values))
```
Comment (zhtmike) [inline verl_omni/trainer/diffusion/ray_diffusion_trainer.py]: pls move these change to a function similar to `compute_xxx_metrics_diffusion`

File: verl_omni/trainer/main_diffusion.py [hunk 4]
```diff
@@ -211,6 +211,10 @@ def run(self, config):
         pprint(OmegaConf.to_container(config, resolve=True))
         OmegaConf.resolve(config)
 
+        from verl_omni.reward_loop.config import resolve_multi_reward_config
+
+        resolve_multi_reward_config(config)
+
```
Comment (zhtmike) [inline verl_omni/trainer/main_diffusion.py]: let us drop this, as mentioned above

**Review by zhtmike — COMMENTED** [review]
Thank you for your PR, overall looks good, please see my few comments

File: verl_omni/reward_loop/config.py (hunk 1 above)
Comment (ruihanglix) [reply verl_omni/reward_loop/config.py]: https://github.com/verl-project/verl-omni/pull/109/commits/e4a5a4c8b74f40afe592059ca74d50875a13e16b

File: verl_omni/trainer/config/reward/reward.yaml (hunk 2 above)
Comment (ruihanglix) [reply verl_omni/trainer/config/reward/reward.yaml]: https://github.com/verl-project/verl-omni/pull/109/commits/e4a5a4c8b74f40afe592059ca74d50875a13e16b

File: verl_omni/trainer/diffusion/ray_diffusion_trainer.py (hunk 3 above)
Comment (ruihanglix) [reply verl_omni/trainer/diffusion/ray_diffusion_trainer.py]: https://github.com/verl-project/verl-omni/pull/109/commits/e4a5a4c8b74f40afe592059ca74d50875a13e16b

File: verl_omni/trainer/main_diffusion.py (hunk 4 above)
Comment (ruihanglix) [reply verl_omni/trainer/main_diffusion.py]: https://github.com/verl-project/verl-omni/pull/109/commits/e4a5a4c8b74f40afe592059ca74d50875a13e16b

**Comment by ruihanglix** [comment]
> Thank you for your PR, overall looks good, please see my few comments

All resolved in https://github.com/verl-project/verl-omni/pull/109/commits/e4a5a4c8b74f40afe592059ca74d50875a13e16b

**Comment by zhtmike** [comment]
Please fix the pre-commit

**Comment by ruihanglix** [comment]
> Please fix the pre-commit

done: https://github.com/verl-project/verl-omni/pull/109/commits/8b2c2086e85c1ab18e382884cff1396b7cd55fe3

**Comment by zhtmike** [comment]
> > Please fix the pre-commit
> 
> done: [8b2c208](https://github.com/verl-project/verl-omni/commit/8b2c2086e85c1ab18e382884cff1396b7cd55fe3)

Seems not fixed.

File: verl_omni/reward_loop/reward_manager/multi.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    params = sig.parameters
+    # Check if the function accepts **kwargs
+    for param in params.values():
+        if param.kind == inspect.Parameter.VAR_KEYWORD:
+            return all_kwargs
+    # Only pass declared parameters
+    return {k: v for k, v in all_kwargs.items() if k in params}
+
+
+class MultiVisualRewardManager(VisualRewardManager):
```
Comment (SamitHuang) [inline verl_omni/reward_loop/reward_manager/multi.py:49]: It's better add a unit test to cover MultiVisualRewardManager module

**Review by SamitHuang — COMMENTED** [review]
LGTM. UT is necessary for this feature.

**Comment by ruihanglix** [comment]
> > > Please fix the pre-commit
> > 
> > 
> > done: [8b2c208](https://github.com/verl-project/verl-omni/commit/8b2c2086e85c1ab18e382884cff1396b7cd55fe3)
> 
> Seems not fixed.

https://github.com/verl-project/verl-omni/pull/109/commits/d06d039ded22d772620e108488e1d3e12b0a5237 @zhtmike 



> LGTM. UT is necessary for this feature.

https://github.com/verl-project/verl-omni/pull/109/commits/a5616868cdac53cad0f520f4e2cf73f5269624cf @SamitHuang

File: verl_omni/reward_loop/reward_manager/multi.py (hunk 5 above)
Comment (ruihanglix) [reply verl_omni/reward_loop/reward_manager/multi.py:49]: https://github.com/verl-project/verl-omni/pull/109/commits/a5616868cdac53cad0f520f4e2cf73f5269624cf

**Comment by zhtmike** [comment]
Please fix the CPU unit test.

**Comment by ruihanglix** [comment]
> Please fix the CPU unit test.

fixed here, https://github.com/verl-project/verl-omni/pull/109/commits/64105ee2d972eab717ac6997b5c71609f9687f72

**Review by zhtmike — APPROVED** [review]
Thank you for your PR!
